Added a very simply way to backup up capabilities.

mattihautameki opened the pull request:
Since capabilities are not backuped with tar I implemented a file based backup using getcap and setcap.This commit was tested on SLES12 and RHEL7. Both use capabilities instead of SUID for ping, arping, etc.

schlomo commented on usr/share/rear/restore/NETFS/default/51_set_capabilities.sh:
```diff
@@ -0,0 +1,11 @@
+				setcap $cap ${TARGET_FS_ROOT}/${file}	
+			done < <(cat $VAR_DIR/recovery/capabilities | sed 's/=//')
+		else
+			Log "setcap binary not found."
```
Did you observe this? Better rely on REQUIRED_PROGS and simply assume that it works. And if not then we should urgently fix REQUIRED_PROGS

gdha replied:
As you added `setcap` and `getcap` to the `REQUIRED_PROGS` array remove the `if which` check as it is unneeded (rest assure the test for REQUIRED_PROGS works fine).

mattihautameki replied:
In my opinion the `REQUIRED_PROGS` in `prep/default/95_check_missing_programs.sh` is not working correctly. Bash is not my favorite thing so maybe I am wrong.  
Due to `MISSING_PROGS=("")` the later `test -n "$MISSING_PROGS"` always detects the first element as null. 
```
[root@cen2 rear]# MISSING_PROGS=("")
[root@cen2 rear]# MISSING_PROGS=( "${MISSING_PROGS[@]}" "missing_bin" )
[root@cen2 rear]# if test -n "$MISSING_PROGS" ; then echo "something is missing" ;fi
```

What works for me is 
```
if [ ${#MISSING_PROGS[@]} -ne 1 ] ; then echo "something is missing" ;fi
```

schlomo commented on usr/share/rear/prep/GNU/Linux/31_include_cap_uitls.sh:
```diff
@@ -0,0 +1,5 @@
+# include utilities needed to set capabilities
+if [ "$BACKUP_CAP" == "y" ] ; then
+    REQUIRED_PROGS=("${REQUIRED_PROGS[@]}" setcap getcap)
+    Log "Tools to set capabilities."
```
If you need to have a log message, maybe call it "Include tools to save/restore capabilites"

mattihautameki replied:
For sure. Thats a careless mistake.

schlomo commented on usr/share/rear/conf/default.conf:
```diff
@@ -949,3 +949,5 @@ export TMPDIR    # the export is required so that mktemp can pickup the variable
 DRLM_MANAGED=n
 
+# Specify if rear should try to backup capabilities (y/n) default (n).
+BACKUP_CAP=n
```
Since the code for BACKUP_CAP is only used for BACKUP=NETFS please rename the variable to something like `NETFS_RESTORE_CAPABILITIES`. That way it will show up if you run `rear dump` and the name of the variable is much more self expalantory.

Also, please move this variable to the NETFS section in `default.conf`

mattihautameki replied:
seems sane

gdha replied:
Please define it as `NETFS_RESTORE_CAPABILITIES=` in the `default.conf` file. To check if a variable is true or not try to use the function `is_true` (defined in lib/global-functions.sh)

schlomo commented on usr/share/rear/backup/NETFS/default/41_save_capabilities.sh:
```diff
@@ -0,0 +1,11 @@
+	if which getcap >/dev/null 2>&1 ; then
+		getcap -r / 2>/dev/null | grep -v $ISO_DIR > $VAR_DIR/recovery/capabilities
+	else
+		Log "getcap binary not found."
```
Let's rely on REQUIRED_PROGS

mattihautameki replied:
I don't see what you point out.

schlomo replied:
I am against having code in different places that checks things that we already check somewhere else. If you put something into REQUIRED_PROGS then we have one place to validate that. All other places should assume that the prog is there without extra checks.

gdha replied:
Use variable `NETFS_RESTORE_CAPABILITIES` instead of `which getcap`. You were right `getcap` was already checked before if it was available via `REQUIRED_PROGS`

schlomo commented on usr/share/rear/backup/NETFS/default/41_save_capabilities.sh:
```diff
@@ -0,0 +1,11 @@
+if [ "$BACKUP_CAP" == "y" ] ; then
+	Log "Saving capabilities to File." 
+	if which getcap >/dev/null 2>&1 ; then
+		getcap -r / 2>/dev/null | grep -v $ISO_DIR > $VAR_DIR/recovery/capabilities
```
IMHO this code should reflect the actual paths that will be part of the backup. If you look at `backup/NETFS/default/50_make_backup.sh` you will see how NETFS backup works.

The include and exclude list of backup paths is provided in `$TMP_DIR/backup-include.txt` and `$TMP_DIR/backup-exclude.txt`

On my Ubuntu desktop `getcap -r /` throws a lot of errors for `/proc`, `/run` etc. which you ignore ATM. I think if you run `getcap` only on the relevant files and directories then you don't need to ignore the errors but can add them to the logfile. In case of a real problem this will help the user to understand why his recovered system does not work properly.

What is the size of the capabilities file? Will it make the boot image too large? Maybe compress with gzip or bzip2 or even put it in OUTPUT_URL instead of on the rescue media. After all, if the backup is not available then the capabilities file is also not needed.

Last, on my desktop with an SSD the `getcap -r /`  takes 3 seconds on the repeated runs and finds 12 items. On a server this might take much longer especially since in most cases the caches are not yet primed. I think that adding a progress bar would be nice to show the user what takes so long.

mattihautameki replied:
In my  `$TMP_DIR/backup-include.txt` all my Filesystems are included. So `/`, `/usr` and so on. So I don't see a way to not do an `getcat -r /`.

At least I dont see a better way. Because without `-r` I would have to use find in conjunction with getcap for all includes and eleminate duplicates afterwards.

schlomo replied:
Yes, the `getcap` tool is flawed and not made for automation. Are there other tools for this task?

Also, what happens if you record capability of a file that is not part of the actual backup? Will `setcap` fail? Or do we silently ignore errors when calling `setcap` to restore the capabilities?

In general, you need to have a clean base line if you want to be able to handle actual errors. If you don't have a clean baseline then you won't be able to distinguish real errors (needing human intervention) from "normal" errors which are safe to ignore.

gdha replied:
Use variable `NETFS_RESTORE_CAPABILITIES` instead of which `getcap`

schlomo commented on usr/share/rear/restore/NETFS/default/51_set_capabilities.sh:
```diff
@@ -0,0 +1,11 @@
+		if which setcap >/dev/null 2>&1 ; then
+			while read file cap ; do
+				setcap $cap ${TARGET_FS_ROOT}/${file}	
+			done < <(cat $VAR_DIR/recovery/capabilities | sed 's/=//')
```
Maybe something like `while read file equal cap` will avoid the sed. The equal sign will go to the `$equal` variable which you can then simply ignore.

schlomo commented on usr/share/rear/restore/NETFS/default/51_set_capabilities.sh:
```diff
@@ -0,0 +1,11 @@
+		Log "Restoring Capabilities."
+		if which setcap >/dev/null 2>&1 ; then
+			while read file cap ; do
+				setcap $cap ${TARGET_FS_ROOT}/${file}	
```
Please quote the file name so that we handle filenames with blanks in them.  In fact, I think that `getcap` and `setcap` will lead to severe problems with bad filenames. Consider this example:

```
$ getcap -r /tmp
/tmp/foo "= foo = cap_net_raw+ep

```

IMHO it is nearly impossible to parse this output in a sane way. I know that this won't affect the system files that you wrote this code for. Maybe we should somehow detect such cases and warn the user that we can't handle it.

schlomo commented:
Hi @mattihautameki,

thanks a lot for this pull request! This is indeed a missing feature and thanks a lot for providing us with a first implementation.

I went through the code and left some detailed comments. IMHO only the following are real blockers:
* rename BACKUP_CAP
* Use quotes for setcap
* parse the capabilities file in a more robust way, e.g. with this example: `while IFS="=" read file cap ; do file="${file% }" cap="${cap# }" ; declare -p file cap ; done <<<'/some file with blanks and " = some_cap,other_cap'`

Kind Regards,
Schlomo

jsmeix commented:
Regarding https://github.com/rear/rear/pull/771#discussion_r52848923 "REQUIRED_PROGS is not working correctly":

Also in my opinion REQUIRED_PROGS is not working correctly, see https://github.com/rear/rear/issues/755#issuecomment-171603580 and subsequent comments.

But I still do not know under what exact circumstances rear should fail with an error if something in REQUIRED_PROGS is missing so that I cannot decide if REQUIRED_PROGS currently works as intended or not.

mattihautameki commented:
Hi! 
I implemented the suggested improvments to the best of my belief.

* Changed BACKUP_CAP to NETFS_RESTORE_CAPABILITIES and moved it to the NETFS Section in `default.conf`.
* Use quotes for filename and capabilities.
* I adapted the MISSING_PROGS check in` 95_check_missing_programs.sh` since the `Error` function is not called even if all binaries from REQUIRED_PROG are missing.

I also moved the block which is collecting the capabilities `41_save_capabilities.sh` to the **resuce** section of **NETFS**. This is because `/var/lib/rear/recovery/capabilities` is not in the ISO from the actual mkbackup-run. When `rear mkbackup` is run a second the file will be included to the ISO because it is already present on the filesystem. 
I am not familiar with the hole code of rear so let me know if there is a better way. 

Kind Regards,
Markus

gdha commented:
@mattihautameki looks OK to me. Thank you for the new (missing) feature.
